local/include/core/http3.h

```diff
@@ -0,0 +1,401 @@
+/** @file
+ * Common data structures and definitions for Proxy Verifier tools.
+ *
+ * Copyright 2021, Verizon Media
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+
+#include "http.h"
+
+#include <chrono>
+#include <deque>
+#include <list>
+#include <memory>
+#include <mutex>
+#include <ngtcp2/ngtcp2.h>
+#include <nghttp3/nghttp3.h>
+#include <openssl/ssl.h>
+#include <random>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "swoc/BufferWriter.h"
+#include "swoc/Errata.h"
+#include "swoc/MemArena.h"
+#include "swoc/swoc_ip.h"
+#include "swoc/TextView.h"
+
+class HttpHeader;
+struct Txn;
+
+namespace swoc
+{
+inline namespace SWOC_VERSION_NS
+{
+namespace bwf
+{
+/** Format wrapper for @c ngtcp2 errors.
+ */
+struct Ngtcp2Error
+{
+  int _e;
+  explicit Ngtcp2Error(int e) : _e(e) { }
+};
+
+/** Format wrapper for @c nghttp3 errors.
+ */
+struct Nghttp3Error
+{
+  int _e;
+  explicit Nghttp3Error(int e) : _e(e) { }
+};
+} // namespace bwf
+
+BufferWriter &bwformat(BufferWriter &w, bwf::Spec const &spec, bwf::Ngtcp2Error const &error);
+BufferWriter &bwformat(BufferWriter &w, bwf::Spec const &spec, bwf::Nghttp3Error const &error);
+} // namespace SWOC_VERSION_NS
+} // namespace swoc
+
+/** Encapsulate the buffer for the QUIC hanshake. */
+class QuicHandshake
+{
+public:
+  QuicHandshake();
+  ~QuicHandshake() = default;
+
+  QuicHandshake(QuicHandshake const &) = delete;
+  QuicHandshake &operator=(QuicHandshake const &) = delete;
+
+public:
+  /** This is the maximum number of bytes we expect to use for the QUIC
+   * handshake.
+   *
+   * This max value is taken from CURL code which has a comment expressing
+   * tentative hope that this should be large enough. There is an assertion in
+   * our (and theirs, I believe) implementation guarding this invariant. If we
+   * trip that, we may need to expand this.
+   */
+  static constexpr size_t max_handshake_size = 4 * 1024;
+
+  /** This contains the storage for the QUIC handshake. */
+  std::vector<char> buf;
```

Comment:
Why isn't this `std::array<char,  max_handshake_size>`? Does it really need to be dynamically sized?

Reply:
Yeah, in an earlier iteration of this I tried std::array. From a memory standpoint it works, but the interface of std::vector is helpful here:
https://github.com/bneradt/proxy-verifier/blob/add_client_side_http3_support/local/src/core/http3.cc#L460

I can insert into the std::vector and get the size of the number of inserted elements from it (not just its capacity). That helps with the implementation of that function. Otherwise I'd have to keep a separate size_t variable and manually keep track of the number of bytes inserted. std::vector offers an interface that does that for me.